src/hypothesis/searchstrategy/regex.py

```diff
@@ -0,0 +1,378 @@
+                    hs.characters(
+                        whitelist_categories=self._categories,
+                        blacklist_characters=self._blacklist_chars,
+                        max_codepoint=max_codepoint,
+                    )
+                )
+            if self._whitelist_chars:
+                strategies.append(
+                    hs.sampled_from(
+                        sorted(self._whitelist_chars - self._blacklist_chars)
+                    )
+                )
+
+        return hs.one_of(*strategies) if strategies else hs.just(u'')
+
+    def add_category(self, category):
+        """Update unicode state to match sre_parse object ``category``."""
+        if category == sre.CATEGORY_DIGIT:
+            self._categories |= UNICODE_DIGIT_CATEGORIES
+        elif category == sre.CATEGORY_NOT_DIGIT:
+            self._categories |= UNICODE_CATEGORIES - UNICODE_DIGIT_CATEGORIES
+        elif category == sre.CATEGORY_SPACE:
+            self._categories |= UNICODE_SPACE_CATEGORIES
+            self._whitelist_chars |= UNICODE_SPACE_CHARS \
+                if self._unicode else SPACE_CHARS
+        elif category == sre.CATEGORY_NOT_SPACE:
+            self._categories |= UNICODE_CATEGORIES - UNICODE_SPACE_CATEGORIES
+            self._blacklist_chars |= UNICODE_SPACE_CHARS \
+                if self._unicode else SPACE_CHARS
+        elif category == sre.CATEGORY_WORD:
+            self._categories |= UNICODE_WORD_CATEGORIES
+            self._whitelist_chars.add(u'_')
+            if HAS_WEIRD_WORD_CHARS and self._unicode:  # pragma: no cover
+                # This code is workaround of weird behavior in
+                # specific Python versions and run only on those versions
+                self._blacklist_chars |= UNICODE_WEIRD_NONWORD_CHARS
+        elif category == sre.CATEGORY_NOT_WORD:
+            self._categories |= UNICODE_CATEGORIES - UNICODE_WORD_CATEGORIES
+            self._blacklist_chars.add(u'_')
+            if HAS_WEIRD_WORD_CHARS and self._unicode:  # pragma: no cover
+                # This code is workaround of weird behavior in
+                # specific Python versions and run only on those versions
+                self._whitelist_chars |= UNICODE_WEIRD_NONWORD_CHARS
+        else:  # pragma: no cover
+            raise InvalidArgument(
+                'Unknown character category: %s' % category
+            )
+
+    def add_chars(self, chars):
+        """Add given chars to char set."""
+        for c in chars:
+            if self._ignorecase:
+                self._whitelist_chars.add(c.lower())
+                self._whitelist_chars.add(c.upper())
+            else:
+                self._whitelist_chars.add(c)
+
+
+def regex_strategy(regex):
+    if not hasattr(regex, 'pattern'):
+        regex = re.compile(regex)
+    return _strategy(
+        sre.parse(regex.pattern),
+        Context(flags=regex.flags),
+        isinstance(regex.pattern, text_type)
+    ).filter(regex.match)
```

Comment:
I think something like the following is probably the correct logic:

```python

def ensure_match(string):
    while True:
        m = regex.match(string)
        assume(m is not None)
        assert m.start() == 0
        if m.end() == len(string):
            return string
        assert m.end() < len(string)
       string = string[:m.end()]
```

Then do `map(ensure_match)` instead of the filter.

This restricts us to the initial prefix that matches, then checks again to see if it still matches. It repeatedly does that, truncating the string down to its initial match until either it finds a prefix where the whole string matches or it produces a non-match.